src/ENCRYPTO_utils/crypto/ecc-pk-crypto.h

```diff
@@ -34,7 +34,16 @@ class ecc_num;
 class ecc_fe;
 class ecc_brickexp;
 
+//The __thread keyword is implementation dependent and therefore
+//it is potentionally dangerous to use this keyword.
+//If the code doesn't compile with your compiler change
+//the MULTI keyword in the cmake configuration to an empty string
+//to disable thread local storage
+#ifdef TRHEADSYNC_ENABLED
+static __thread std::mutex relic_mutex;
+#else
 static std::mutex relic_mutex;
+#endif

```

Comment:
So if multithreading via pthreads is enabled in Relic, then you give every thread its own mutex in thread local storage. Hence, every thread locks its own mutex to access its own relic context object (also in TLS). I do not see the need for this mutex then.

Reply:
I've already considered it to disable the mutex completely if the preprocessor flag is defined (and this was done in a previous implementation of mine). However after some considerations I decided against it since the method only enables thread safety for a single thread method; that is, if a user decides to use pthread but then also uses openmp concurrently thread safety cannot be guranteed then, and it is not obvious to see for a user if he doesn't study the code.

I rather prefer to implement concurrency conserative since errors are often easy to make and hard to find later. Besides that there speaks nothing against your suggestion.